### ISSUE TITLE
Allow testing external PRs, but without double builds

### ISSUE THREAD

**Issue by jfirebaugh** [travis-ci/travis-ci#6612]
I want to configure Travis such that:

* A push to any branch triggers exactly one build, whether there is a PR for that branch or not. (I.e. if there is a PR, _do not_ trigger a build of a synthesized merge commit.)
* Opening a new _external_ PR (or pushing to an existing one) also triggers exactly one build.

As far as I can tell, there's no way to do this with the current build options. If "Build pull requests" is not enabled, external PRs don't get built. If it is enabled, internal PRs trigger two builds on every push. The suggestion [here](http://stackoverflow.com/questions/31882306/how-to-configure-travis-ci-to-build-pull-requests-merges-to-master-w-o-redunda) of using a `branches:` whitelist doesn't satisfy the first criterion -- a branch that isn't whitelisted and doesn't have a PR won't get built.

**Comment by aantron** [comment]
I need exactly this. Not having this wastes both my time and build worker resources.

I generally want branches tested, not merges. So, I don't want PR testing for PRs from branches in the same repo. However, if there is a PR incoming from outside the repo, that PR should be tested using the repo's build setup. There appears to be no way to configure this currently.

**Comment by oshalygin** [comment]
This is pretty useful.  Anyone have a clean way of accomplishing this?  Maybe a temporary workaround?

**Comment by ruslo** [comment]
> This is pretty useful. Anyone have a clean way of accomplishing this? Maybe a temporary workaround?

I'm using this one:
* http://stackoverflow.com/a/41715756/2288008

**Comment by oshalygin** [comment]
> I'm using this one:

http://stackoverflow.com/a/41715756/2288008

Not bad, thanks for sharing!!

**Comment by jfirebaugh** [comment]
While clever, renaming the branch when opening a pull-request is still a cumbersome workaround, in my opinion. And renaming also triggers an unnecessary build, the very thing I'd like to avoid.

**Comment by ruslo** [comment]
> And renaming also triggers an unnecessary build, the very thing I'd like to avoid

It will not because the name of the branch match exclude pattern.

**Comment by jfirebaugh** [comment]
Right, but doesn't opening the PR trigger a PR build, even if that commit SHA was previously built under the old branch name?

**Comment by ruslo** [comment]
> Right, but doesn't opening the PR trigger a PR build, even if that commit SHA was previously built under the old branch name?

Yes, if you push it to remote. The idea is that if you know that feature branch is ready to be merged to target branch then don't push `wip.feature`, you can create `pr.feature` **locally** and push it.

**Comment by jfirebaugh** [comment]
I've moved most projects I'm actively working on to CircleCI (due to this and other issues), so I don't personally mind if it's closed. But there seem to be quite a few other participants who would still like it to be addressed.

**Comment by svenevs** [comment]
FWIW AppVeyor also supports this [via `skip_branch_with_pr: true` ](https://github.com/appveyor/ci/issues/882), it makes me sad to think that Travis is losing people over something like this...the GitHub API allows you to find this information (see the sample `GET` in the first post linked).

**Comment by notmgsk** [comment]
Bump. Still an issue.
